database/src/mdbx/database.rs

```diff
@@ -1,171 +1,149 @@
-use std::{borrow::Cow, fs, path::Path, sync::Arc};
+use std::{any::TypeId, fs, ops::Range, path::Path, sync::Arc};
 
 use libmdbx::NoWriteMap;
-use log::info;
+use log::{debug, info};
+use tempfile::TempDir;
 
 use super::{MdbxReadTransaction, MdbxWriteTransaction};
-use crate::{traits::Database, DatabaseProxy, Error, TableFlags};
+use crate::{
+    traits::{AsDatabaseBytes, Database, DupTable, RegularTable, Table},
+    Error,
+};
+
+const GIGABYTE: usize = 1024 * 1024 * 1024;
+const TERABYTE: usize = GIGABYTE * 1024;
+
+/// Database config options.
+pub struct DatabaseConfig {
+    pub max_tables: Option<u64>,
+    pub max_readers: Option<u32>,
+    pub no_rdahead: bool,
+    pub size: Option<Range<isize>>,
+    pub coalesce: bool,
+    pub growth_step: Option<isize>,
+    pub shrink_threshold: Option<isize>,
+}
+
+impl Default for DatabaseConfig {
+    fn default() -> Self {
+        // Use more sensible defaults.
+        DatabaseConfig {
+            max_tables: None,
+            max_readers: None,
+            no_rdahead: true,
+            // Default max database size: 2TB
+            size: Some(0..(2 * TERABYTE as isize)),
+            coalesce: false,
+            // Default growth step: 4GB
+            growth_step: Some(4 * GIGABYTE as isize),
+            shrink_threshold: None,
+        }
+    }
+}
 
-pub(super) type DbKvPair<'a> = (Cow<'a, [u8]>, Cow<'a, [u8]>);
+impl From<DatabaseConfig> for libmdbx::DatabaseOptions {
+    fn from(value: DatabaseConfig) -> Self {
+        libmdbx::DatabaseOptions {
+            max_tables: value.max_tables,
+            max_readers: value.max_readers,
+            no_rdahead: value.no_rdahead,
+            mode: libmdbx::Mode::ReadWrite(libmdbx::ReadWriteOptions {
+                sync_mode: libmdbx::SyncMode::Durable, // default anyway
```

Comment:
I don't really see the value of the comment
```suggestion
                sync_mode: libmdbx::SyncMode::Durable,
```